File: src/radar.py
import math
import sys
import pandas as pd

# Loading data 
df = pd.read_csv("data/pl_teams.csv")
# Clean the column possession which represents in %
df['Possession'] = df['Possession'].str.rstrip('%').astype(float)

labels = ["Interceptions", "Tackles", "Passes", "Possession", "Goals", "Shots", "Fouls"]

team_data = {
    row['Teams']: [row[label] for label in labels] 
    for _, row in df.iterrows()
}

team_colors = {
    row['Teams']: row['Color'].replace(',', '') 
    for _, row in df.iterrows()
}

# multiply by 1.1 to extend the maximum
max_vals = [df[label].max() * 1.1 for label in labels]

def get_jgraph(team_names):
    if not(team_names):
        # include all teams
        team_names = team_data.keys()
    for name in team_names:
        if name not in team_data:
            print(f"(* Error: Team {name} not found *)")
            return

    center_x, center_y = 5, 5
    max_radius = 4
    num_axes = len(labels)

    print("newgraph")
    print("xaxis min 0 max 10 nodraw")
    print("yaxis min 0 max 10 nodraw")

    # Legend configuration
    print("legend top")

    # change as needed
    web_distance = [0.2, 0.4, 0.6, 0.8, 1.0]
    scale = 1.5
    # Background web
    for level in web_distance:
        print(f"newcurve linetype solid gray .8 poly pfill -1 pts", end="")
        for i in range(num_axes):
            angle = (i * 2 * math.pi / num_axes) + (math.pi / 2)
            px = center_x + (scale * level * max_radius * math.cos(angle))
            py = center_y + (scale * level * max_radius * math.sin(angle))
            print(f" {px:.2f} {py:.2f}", end="")
        print()

    #draw 7 lines from center_x, center_y to last polygon
    for i in range(num_axes):
        angle = (i * 2 * math.pi / num_axes) + (math.pi / 2)
        px = center_x + (scale * 1.0 * max_radius * math.cos(angle))
        py = center_y + (scale * 1.0 * max_radius * math.sin(angle))
        print(f"newcurve linetype solid gray .8 pts {center_x} {center_y} {px:.2f} {py:.2f}")

    # label each levels with a % level
    for level in web_distance:
        angle = (0 * 2 * math.pi / num_axes) + (math.pi / 2)
        lx = center_x + (scale * level * max_radius * math.cos(angle))
        ly = center_y + (scale * level * max_radius * math.sin(angle))
        # Offset the label slightly to the right for visibility
        print(f"newstring hjl vjc fontsize 8 x {lx+0.1:.2f} y {ly:.2f} : {int(level*100)}%")

    # Draw the team polygons with legends
    for index, name in enumerate(team_names):
        stats = team_data[name]
        color = team_colors.get(name, "0.5 0.5 0.5")
        
        # Polygon with label for the legend
        print(f"newline poly ", end="")
        print(f"color {color} linethickness 5 pfill -1 label : {name} \n pts", end="")
        
        points = []
        for i in range(num_axes):
            normalized = stats[i] / max_vals[i]
            angle = (i * 2 * math.pi / num_axes) + (math.pi / 2)
            px = center_x + (scale * normalized * max_radius * math.cos(angle))
            py = center_y + (scale * normalized * max_radius * math.sin(angle))
            points.append((px, py))
            print(f" {px:.2f} {py:.2f}", end="")
        print()


    # Labels
    for i in range(num_axes):
        angle = (i * 2 * math.pi / num_axes) + (math.pi / 2)
        lx = center_x + (1.1 * scale * max_radius * math.cos(angle))
        ly = center_y + (1.1 * scale * max_radius * math.sin(angle))
        print(f"newstring hjc vjc font Helvetica fontsize 10 x {lx:.2f} y {ly:.2f} : {labels[i]}")

    # Generating Title
    if len(team_names) == 2: title = " vs ".join(team_names) + " Stats"
    elif len(team_names) == 1: title = team_names[0] + " Playing Stats"
    else: title = "Premier League Teams Stats"
    print(f"newstring hjc vjc x 5 y -5 font Helvetica-Bold fontsize 10 : {title}")

if __name__ == "__main__":
    get_jgraph(sys.argv[1:])

File: src/radar.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <vector>
#include <string>
#include <map>
#include <cmath>
#include <iomanip>
#include <algorithm>
#include <random>

using namespace std;

struct Team {
    string name;
    vector<double> stats;
};

// cleaning text strings
string trim(string s, char extra = ' ') {
    s.erase(remove(s.begin(), s.end(), extra), s.end());
    return s;
}

// metrics
const vector<string> labels = {"Interceptions", "Tackles", "Passes", "Possession", "Goals", "Shots", "Fouls"};
const double center_x = 5.0, center_y = 5.0, max_radius = 4.0;
const double scale = 1.5; // scale of the graph

// function to provide x and y, based on centre and radius
void get_coords(int index, double radius, double &x, double &y) {
    int num_axes = labels.size();
    double angle = (index * 2 * M_PI / num_axes) + (M_PI / 2);
    x = center_x + (scale * radius * max_radius * cos(angle));
    y = center_y + (scale * radius * max_radius * sin(angle));
}

// get random color for each team polygon (got from google search)
string get_random_color() {
    static mt19937 gen(time(0));
    uniform_real_distribution<> dis(0.0, 1.0);
    stringstream ss;
    ss << fixed << setprecision(2) << dis(gen) << " " << dis(gen) << " " << dis(gen);
    return ss.str();
}

map<string, Team> load_data(const string& filename, vector<double>& max_vals) {
    ifstream file(filename);
    map<string, Team> team_map;
    string line, word;

    if (!file.is_open()) return team_map;

    // this skips the header
    getline(file, line);
    max_vals.assign(labels.size(), 0.0);

    while (getline(file, line)) {
        stringstream ss(line);
        vector<string> row;
        while (getline(ss, word, ',')) row.push_back(word);
        
        // there needs to be 8 colums (team name + 7 metrics)
        if (row.size() < 8) continue;

        Team t;
        t.name = row[0];

        for (int i = 0; i < 7; ++i) {
            string val_str = row[i + 1];
            // the possession contains the value in %, remove '%' sign
            val_str.erase(remove(val_str.begin(), val_str.end(), '%'), val_str.end());
            double val = stod(val_str);
            t.stats.push_back(val);
            if (val > max_vals[i]) max_vals[i] = val;
        }
        team_map[t.name] = t;
    }
    // scale max vals by 1.1 just to stretch the radar
    for (double &m : max_vals) m *= 1.1;
    return team_map;
}

void render_jgraph(const vector<string>& team_names, map<string, Team>& team_map, const vector<double>& max_vals) {
    vector<double> web_distance = {0.2, 0.4, 0.6, 0.8, 1.0};
    int num_axes = labels.size();

    cout << fixed << setprecision(2);
    cout << "newgraph\nxaxis min 0 max 10 nodraw\nyaxis min 0 max 10 nodraw\nlegend top\n";

    // background web
    for (double level : web_distance) {
        cout << "newcurve linetype solid gray .8 poly pfill -1 pts";
        for (int i = 0; i < num_axes; ++i) {
            double px, py;
            get_coords(i, level, px, py);
            cout << " " << px << " " << py;
        }
        cout << endl;
    }
    // the lines from center to the last web
    for (int i = 0; i < num_axes; ++i) {
        double px, py;
        get_coords(i, 1.0, px, py);
        cout << "newcurve linetype solid gray .8 pts " << center_x << " " << center_y << " " << px << " " << py << endl;
    }

    // percentage labels on the web
    for (double level : web_distance) {
        double lx, ly;
        get_coords(0, level, lx, ly);
        cout << "newstring hjl vjc fontsize 8 x " << (lx + 0.1) << " y " << ly << " : " << (int)(level * 100) << "%\n";
    }

    // constructing polygons for each team
    for (const string& name : team_names) {
        if (team_map.find(name) == team_map.end()) continue;
        Team& t = team_map[name];
        cout << "newline poly color " << get_random_color() << " linethickness 5 pfill -1 label : " << name << "\npts";
        for (int i = 0; i < num_axes; ++i) {
            double px, py;
            get_coords(i, (t.stats[i] / max_vals[i]), px, py);
            cout << " " << px << " " << py;
        }
        cout << endl;
    }

    // the labels representing metrics at the end of each axis
    for (int i = 0; i < num_axes; ++i) {
        double lx, ly;
        get_coords(i, 1.15, lx, ly);
        cout << "newstring hjc vjc font Helvetica fontsize 10 x " << lx << " y " << ly << " : " << labels[i] << endl;
    }
}


int main(int argc, char* argv[]) {
    vector<double> max_vals;
    map<string, Team> team_map = load_data("data/pl_teams.csv", max_vals);

    if (team_map.empty()) {
        cerr << "(* Error: No data loaded *)" << endl;
        return 1;
    }

    // if team name is specified in arg use it to create a file, if not use all files
    vector<string> requested_teams;
    if (argc > 1) {
        for (int i = 1; i < argc; ++i) requested_teams.push_back(argv[i]);
    } else {
        for (auto const& [name, data] : team_map) requested_teams.push_back(name);
    }

    render_jgraph(requested_teams, team_map, max_vals);

    return 0;
}